File: ruby/5kyu/paginationhelper.rb
=begin
Codewars. 22/07/20. 'PaginationHelper'. 5kyu. Here we create a Pagination Helper class, which is a utility class used to query paging
information related to an array. The class takes in an array of items and an integer denoting how many items to each page. It contains
various methods. Here is the solution I developed to solve the challenge.
1) When defining a class in Ruby, we use the class keyword, and the convention is to name the class in PascalCase. PascalCase is different
   from camelCase.
2) We create our initialize method, which will take in the properties of our class object, the array of items and the items per page.
3) Inside our initialize method, we store our class properties in instance variables, which makes these inputs accessible in all the
   methods of our class.
4) Our first method is item_count, it takes in no arguments and simply returns the amount in items in our collections array.
5) We then have page_count, this method returns how many pages our array has. We find this by simply dividing our array length by our
   items per page and if there is remainder, we have another page with less than @ipp items so we add 1. For example, if our array has 8
   items and we have 3 items per page, 3 goes into 8 twice, then we have 2 items remaining, so we have 3 pages.
6) page_item_count find how many items are in a given page that we pass to it as an argument. The argument is the page_index, so if we 
   have 3 pages we will pass 0, 1 and 2 are valid page indices to pass.
7) Thus we use the each_slice method to slice the array into seperate pages, each page will have @ipp amount of items, but if @arr.length
   divided by @ipp produces remainder, the last slice will have the remainder amount of items. We simply use bracket notation and pass in
   page_index then take its length, if we get an error, that's because we passed an invalid index, so in that case we use rescue to
   return -1 if we get an error.
8) Our final method is page_index, which takes an item index for the array, and returns what page that item is on.
9) First we handle our error checking, so if item_index is greater than or equal to the length of the array, or if item_index is a negative
   number, we return -1.
10) Otherwise, we divide item_index by @ipp, which gives us the page the item is on.
11) When creating a class object instance, we use ClassName.new(property1, property2...).
=end

class PaginationHelper
  def initialize(collection, items_per_page)
    @arr = collection
    @ipp = items_per_page
  end
  
  def item_count
    @arr.length
  end
  
  def page_count
    (@arr.length % @ipp != 0 ? 1 : 0) + @arr.length / @ipp
  end
  
  def page_item_count(page_index)
    Array(@arr.each_slice(@ipp))[page_index].length rescue -1
  end
  
  def page_index(item_index)
    return -1 if item_index >= @arr.length || item_index.negative?
    item_index / @ipp 
  end
end

File: javascript/5kyu/paginationhelper.js
/*
Codewars. 03/09/20. 'PaginationHelper'. 5kyu. Here we create a utility class helpful for querying paging information related to an array.
The object we create contains two properties, collection which actuals as individual items, and itemsPerPage, which is how many items are on
each page. The actual contents of the array are irrelevant. Here is the solution I developed to solve the challenge.
1) We first create our object constructor, which defines the properties of our object. The name of our object - PaginationHelper - is
   capitalized, as is the convention with object names.
2) 'this' refers to the instance of the object, thus we set this.arr to equal the array we pass in when we create an instance of our
   object and we set this.ipp to equal the amount of items per page that we specify.
3) Here we are using the prototype model - JavaScript is a prototype based language, not a class based language - and we can place functions
   inside the object constructor, but it is not recommended to do so. Instead, we add methods to the object using the prototype property.
4) The prototype property of any object can be accessed using ObjectName.prototype e.g. PaginationHelper.prototype, Number.prototype,
   Array.prototype etc.
5) Thus we add an itemCount function to the object, the itemCount function takes no argument and as with all other object methods, it
   is called directly on the object book1.itemCount(). The itemCount function returns the length of the array i.e. how many total items
   there are in the array.
6) In our pageCount function, which also takes no parameters, we return how many pages the array has. This is done by diving the array
   length by the items per page, then rounding up the result if decimals are producted. For example, if our array is [1,2,3,4,5,6,7,8,9]
   and this.ipp = 4, we have 3 pages. 9 / 4 = 2.25, then rounded up = 3. If our array is ['a','b','c','d','e','f','g','h','i','j','k']
   and this.ipp = 3, we have 4 pages. 11 / 3 = 3.67, then rounded up = 4.
7) Our pageItemCount function takes a pageIndex (0 based) as its argument and returns how many items are on this page. 
8) First we call the pageCount function on our object instance (using this) and store the result in pCount. If pageIndex + 1 (to convert
   the index into an ordinal page number) is greater than the amount of pages in the object, we return -1.
9) Now we find out how many pages in the array are full pages. We divide the array length by ipp and then round down the result, which will
   exclude non-full pages.
10) If the pageIndex + 1 (to convert pageIndex to pageNumber) is less than or equal to fullPages, we return the items per page, otherwise
    we return the remainder of dividing array length by ipp. Remember, the remainder will always be less than the array length and can only
    occur at the end of the array, if it does occur.
11) Our final function is a pageIndex function, which takes an itemIndex (0 based) and returns what page that item is on.
12) First we handle the error checking, if the item index is greater than the highest array index, or less than 0, we return -1.
13) Then we simply divide the itemIndex by ipp and floor the result. For example, if our array is [1,2,3,4,5,6,7,8,9] and this.ipp = 4,
    itemIndex(8) will involve doing 8 / 4 = 2. It is correct that 9 is on page 3, which is pageIndex 2.
14) The syntax for creating an object instance is const book = new PaginationHelper([1,2,3,4,5,6,7,8,9], 4).
*/

function PaginationHelper(collection, itemsPerPage) {
  this.arr = collection;
  this.ipp = itemsPerPage;
}

PaginationHelper.prototype.itemCount = function() {
  return this.arr.length;
};

PaginationHelper.prototype.pageCount = function() {
  return Math.ceil(this.arr.length / this.ipp);
};

PaginationHelper.prototype.pageItemCount = function(pageIndex) {
  let pCount = this.pageCount();
  if (pageIndex + 1 > pCount) {return -1;}
  let fullPages = Math.floor(this.arr.length / this.ipp);
  return (pageIndex + 1) <= fullPages ? this.ipp : this.arr.length % this.ipp;
};

PaginationHelper.prototype.pageIndex = function(itemIndex) {
  if (itemIndex > this.arr.length - 1 || itemIndex < 0) {return -1;}
  return Math.floor(itemIndex / this.ipp);
};

/*
Here we create the same object using JavaScript's class syntax. Classes were introduced in ES6 to primarily provide syntactic sugar over
JavaScript's existing prototype model. Under the hood, JS classes do exactly the same thing as prototypes and in fact are using the
prototype model. As stated before, classes do not actually exist in JavaScript, as JavaScript is a prototype-based language, classes merely
provide us with an alternative syntax.
1) Like with the prototype model, the object name is capitalized, but unlike the prototype model, an opening and closing curly brace wrap
   around all functions of the class.
2) The class syntax using a constructor function, where the properties of the object are defined.
3) Our class functions do not require a function keyword.
4) Creating a class instance is the same as in the prototype model i.e. const book = new PaginationHelperC([1,2,3,4,5,6,7,8,9], 4).
*/

class PaginationHelperC {
  constructor(collection, itemsPerPage) {
    this.arr = collection;
    this.ipp = itemsPerPage;
  }

  itemCount() {
    return this.arr.length;
  }

  pageCount() {
    return Math.ceil(this.arr.length / this.ipp);
  }

  pageItemCount(pageIndex) {
    let pCount = this.pageCount();
    if (pageIndex + 1 > pCount) {return -1;}
    let fullPages = Math.floor(this.arr.length / this.ipp);
    return pageIndex + 1 <= fullPages ? this.ipp : this.arr.length % this.ipp;
  }

  pageIndex(itemIndex) {
    if (itemIndex < 0 || itemIndex > this.arr.length - 1) {return -1;}
    return Math.floor(itemIndex / this.ipp);
  }
}